assets/javascripts/discourse/components/chat-live-pane.js

```diff
@@ -905,34 +925,69 @@ export default Component.extend({
 
         // Make sure new messages have come in. Do not keep pinging server with read updates
         // if no new messages came in since last read update was sent.
-        if (this._floatOpenAndFocused() && hasUnreadMessage) {
-          this.set("lastSendReadMessageId", messageId);
-          ajax(`/chat/${this.chatChannel.id}/read/${messageId}.json`, {
+        if (this._floatOpenAndFocused() && hasUnreadMessages) {
+          this.set("lastSendReadMessageId", latestUnreadMsgId);
+          ajax(`/chat/${this.chatChannel.id}/read/${latestUnreadMsgId}.json`, {
             method: "PUT",
-          }).catch(() => {
-            this._stopLastReadRunner();
           });
         }
-
-        this._updateLastReadMessage();
       },
       wait
     );
   },
 
-  _floatOpenAndFocused() {
-    return userPresent() && this.expanded && !this.floatHidden;
-  },
+  // Recursive binary search to find an unread message in the viewport.
+  _findNextUnreadMessage(start, end, lastReadMessageId) {
+    if (end < start) {
+      return this.messages[end]?.id;
+    }
 
-  _startLastReadRunner() {
-    if (!isTesting()) {
-      next(this, () => {
-        this._updateLastReadMessage(0);
-      });
+    const candidateIdx = Math.round((start + end) / 2);
+    const unreadCandidateId = this.messages[candidateIdx]?.id;
+
+    if (!unreadCandidateId) {
+      return this.messages[end]?.id;
+    }
+
+    if (unreadCandidateId > lastReadMessageId) {
+      const messageElem = document.querySelector(
+        `.chat-message-container[data-id="${unreadCandidateId}"]`
+      );
+
+      if (this._messageVisible(messageElem)) {
+        return unreadCandidateId;
+      } else {
+        // Right half but not visible yet. Keep searching.
+        return this._findNextUnreadMessage(
+          start,
+          candidateIdx - 1,
+          lastReadMessageId
+        );
+      }
+    } else {
+      // Wrong half. Look in the rest of the array
+      return this._findNextUnreadMessage(
+        candidateIdx + 1,
+        end,
+        lastReadMessageId
+      );
     }
   },
 
-  _stopLastReadRunner() {
+  _messageVisible(message) {
```

Comment:
How about we try something using IntersectionObserver?

Reply:
What do you have in mind? I don't immediately see how the observer would fit here. We don't know which of the unread messages is visible.

Reply:
I implemented the observer, but added a class instead of a data attribute. I forgot about that part of your suggestion. Is there any difference between class vs data attr?